crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1339,96 +1434,201 @@ impl<'db> Type<'db> {
                 "__bool__" => {
                     // TODO should be `Callable[[], Literal[True/False]]`
                     Symbol::todo("`__bool__` for `AlwaysTruthy`/`AlwaysFalsy` Type variants")
                 }
-                _ => Type::object(db).member(db, name),
+                _ => Type::object(db).static_member(db, name),
+            },
+        }
+    }
+
+    /// Call the `__get__(instance, owner)` method on a type, if it exists.
+    fn try_call_dunder_get(
+        &self,
+        db: &'db dyn Db,
+        instance: Option<Type<'db>>,
+        owner: Type<'db>,
+    ) -> Option<Type<'db>> {
+        // TODO: Handle possible-unboundness and errors from `__get__` calls.
+
+        match self {
+            Type::Union(union) => {
+                let mut builder = UnionBuilder::new(db);
+                for elem in union.elements(db) {
+                    let ty = if let Some(result) = elem.try_call_dunder_get(db, instance, owner) {
+                        result
+                    } else {
+                        *elem
+                    };
+                    builder = builder.add(ty);
+                }
+                Some(builder.build())
+            }
+            Type::Intersection(intersection) => {
+                if !intersection.negative(db).is_empty() {
+                    return Some(todo_type!(
+                        "try_call_dunder_get: intersections with negative contributions"
+                    ));
+                }
+
+                let mut builder = IntersectionBuilder::new(db);
+                for elem in intersection.positive(db) {
+                    let ty = if let Some(result) = elem.try_call_dunder_get(db, instance, owner) {
+                        result
+                    } else {
+                        *elem
+                    };
+                    builder = builder.add_positive(ty);
+                }
+                Some(builder.build())
+            }
+            _ => {
+                // TODO:
+                // - Handle possible-unboundness of `__get__` method
+                // - Handle errors while calling `__get__`
+                //
+                // There are existing tests for both cases in `descriptor_protocol.md`.
+
+                self.member(db, "__get__")
+                    .ignore_possibly_unbound()?
+                    .call(
+                        db,
+                        &CallArguments::positional([instance.unwrap_or(Type::none(db)), owner]),
+                    )
+                    .map(|outcome| Some(outcome.return_type(db)))
+                    .unwrap_or(None)
+            }
+        }
+    }
+
+    /// Access an attribute of this type, potentially invoking the descriptor protocol.
+    /// Corresponds to `getattr(<object of type 'self'>, name)`.
+    ///
+    /// See also: [`Type::static_member`]
+    #[must_use]
+    pub(crate) fn member(&self, db: &'db dyn Db, name: &str) -> Symbol<'db> {
```

Comment:
Could we add a `TODO` here that it should return a `Result` instead of the `Symbol` directly.